src/hdr/decoder.rs

```diff
@@ -323,46 +327,80 @@ impl<R: BufRead> HDRDecoder<R> {
         let uszwidth = self.width as usize;
 
         let pixel_count = self.width as usize * self.height as usize;
+        // Did not overflow, all chunks are full sized.
+        assert!(pixel_count % uszwidth == 0);
+        assert!(pixel_count / uszwidth == self.height as usize);
+
         let mut ret = Vec::with_capacity(pixel_count);
-        unsafe {
-            // RGBE8Pixel doesn't implement Drop, so it's Ok to drop half-initialized ret
-            ret.set_len(pixel_count);
-        } // ret contains uninitialized data, so now it's my responsibility to return fully initialized ret
+
+        // A unique pointer to an uninitialized portion of a slice.  This emulates having sent a
+        // `&mut [T]`, which would be possible, but allows the target region to still be
+        // uninitialized.
+        struct UninitSlice<U: Send>(ptr::NonNull<U>, usize);
+
+        unsafe impl<U: Send> Send for UninitSlice<U> { }
 
         {
-            let chunks_iter = ret.chunks_mut(uszwidth);
+            let chunks_base: *mut T = ret.as_mut_ptr();
             let mut pool = Pool::new(8); //
 
             try!(pool.scoped(|scope| {
-                for chunk in chunks_iter {
+                for chunk_index in 0..self.height as usize {
                     let mut buf = Vec::<RGBE8Pixel>::with_capacity(uszwidth);
                     unsafe {
+                        // SAFETY: allocation has exactly that size.
+                        ptr::write_bytes(buf.as_mut_ptr(), 0u8, uszwidth);
+                        // SAFETY: zeroed is valid for RGBE8Pixel and we initialized just above.
                         buf.set_len(uszwidth);
                     }
                     try!(read_scanline(&mut self.r, &mut buf[..]));
+
                     let f = &f;
+                    // Each chunk is a unique slice of exactly `uszwidth` uninitialized elements.
+                    let chunk = unsafe {
+                        // SAFETY: within the allocation as per `pixel_count` assertions.This is
+                        // in-bounds since the `Vec` has enough capacity and is never re-allocated.
+                        let start: *mut T = chunks_base.add(chunk_index * uszwidth);
+                        // INVARIANT: unique since all slices are spaced `uszwidth` elements and
+                        // also exactly as large.
+                        UninitSlice(ptr::NonNull::<T>::new_unchecked(start), uszwidth)
```

Comment:
With `NonNull::new(start).unwrap()` this wouldn't need to be in an unsafe block.

Reply:
Since it doesn't come with any more proof obligations than the `add` in the line above, I didn't think it would make any difference but simply make the intent even clearer. Note that the null pointer would not be inbounds and thus can not be the result of the computation of `start`.